src/NuGet.Clients/VsExtension/VisualStudioAccountProvider.cs

```diff
@@ -0,0 +1,390 @@
+                }
+
+                result = await provider.AcquireAdalTokenAsync(
+                    resource: VsoEndpointResource,
+                    tenantId: tenantId,
+                    identitifer: new UserIdentifier(uniqueId, UserIdentifierType.UniqueId),
+                    parentWindowHandle: parent,
+                    accountKeyForReAuthentication: account.UserAccount,
+                    prompt: shouldPrompt,
+                    cancellationToken: cancellationToken).ConfigureAwait(false);
+            });
+
+            var aadcred = new VssAadCredential(new VssAadToken(result));
+
+            // create the session token
+            var connection = new VssConnection(AccountManager.VsoEndpoint, aadcred);
+            var delegatedClient = connection.GetClient<DelegatedAuthorizationHttpClient>();
+
+            SessionToken sessionToken = null;
+            // N.B. uncomment scope once task #418464 is done
+            // [Expose packaging token scopes in the UI (also add PUT to vso.packaging_write)]
+            sessionToken = await delegatedClient.CreateSessionToken( /*scope: "vso.packaging_write"*/)
+                .ConfigureAwait(false);
+
+            var cred = new NetworkCredential
+            {
+                UserName = account.UserAccount.DisplayInfo.UserName,
+                Password = sessionToken.Token
+            };
+
+            return cred;
+        }
+
+        // Internal so we can mock.  Need to call this a lot and 
+        internal virtual TenantInformation FindTenantInAccount(VsUserAccount account, string tenantId,
+            VSAccountProvider provider)
+        {
+            var tenantsInScope = provider.GetTenantsInScope(account);
+            return tenantsInScope.FirstOrDefault(tenant => tenant.TenantId == tenantId);
+        }
+
+        // Logic to query web.  This will be mocked out in unit tests
+        internal virtual async Task<string> LookupTenant(Uri uri, IWebProxy proxy,
+            CancellationToken cancellationToken)
+        {
+            string tenantId;
+            //  we assume the call will be access denied (or the provider shouldn't have been called)
+            //  so calling the URI shouldn't be too expensive.
+            var req = WebRequest.Create(uri);
+            if (proxy != null)
+            {
+                req.Proxy = proxy;
+            }
+
+            cancellationToken.ThrowIfCancellationRequested();
+
+            try
+            {
+                using (var response = await req.GetResponseAsync().ConfigureAwait(false))
+                {
+                    tenantId = response.Headers[VssResourceTenant];
+                }
+            }
+            catch (WebException ex)
+            {
+                //  unauthorized is thrown as an exception from GetResponse so we have to pull
+                //  the headers out of the exception as well.
+                tenantId = ex.Response.Headers[VssResourceTenant];
```

Comment:
That has been validated and the behavior is somewhat documented if you look at NameValueCollection (https://msdn.microsoft.com/en-us/library/system.collections.specialized.namevaluecollection(v=vs.110).aspx).  The caution at the bottom states that null can be returned for both not found and for null value off  get call.

If this is still a concern we can wrap this in a check to see if the AllKeys list contains the value before we access it.